File: src/p0405/cpp/solution.cpp
#include <iostream>
#include <string>

using namespace std;


class Solution {
public:
    string toHex(uint32_t num) {
        string hex;
        bool flag = false;
        Byte *bs = (Byte*)&num;
        for (int i = 3; i >= 0; i--) {
            if (bs[i].y || flag) {
                hex.push_back(map[bs[i].y]);
                flag = true;
            }
            if (bs[i].x || flag || i == 0) {
                hex.push_back(map[bs[i].x]);
                flag = true;
            }
        }
        return hex;
    }

private:
    struct Byte { unsigned char x: 4; unsigned char y: 4; };
    static constexpr char map[] = {'0', '1', '2', '3', '4', '5', '6', '7',
                                   '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
};

File: src/p0405/java/solution.java
class Solution {
    public String toHex(int num) {
        StringBuilder hex = new StringBuilder();
        for (int s = 28; s >= 0; s -= 4) {
            int b = (num >>> s) & 0xf;
            if (b != 0 || hex.length() > 0 || s == 0) {
                hex.append(map.charAt(b));
            }
        }
        return hex.toString();
    }
    private static final String map = "0123456789abcdef";
}
